### PR TITLE
add a new module to configure multicast igmp and enable it.

### PR DESCRIPTION
**Pull request by yuandongx**
##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py
### test case
```
  - name: config global igmp enable
    ce_multicast_igmp_enable:
      aftype: v4
      features: 'global'
      state: present


```
### test log info 
```

changed: [10.190.121.125] => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python"
    },
    "changed": true,
    "end_state": {
        "igmp_info": [
            {
                "addrFamily": "ipv4unicast"
            }
        ]
    },
    "existing": {
        "igmp_info": []
    },
    "invocation": {
        "module_args": {
            "aftype": "v4",
            "features": "global",
            "host": "10.190.121.125",
            "igmp": "false"
            "proxy": "false",
            "ssh_keyfile": null,
            "state": "present",
            "timeout": null,
            "transport": "cli",
            "use_ssl": null,
            "username": "huawei",
            "validate_certs": null,
            "version": 2,
            "vlan_id": null
        }
    },
    "proposed": {
        "addrFamily": "ipv4unicast",
        "features": "global",
        "state": "present"
    },
    "updates": [
        "igmp snooping enable"
    ]
}
META: ran handlers
META: ran handlers

PLAY RECAP ******************************************************************************************************************************************************
10.190.121.125             : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

[~8850_130.12-Vlanif3111]di cu | in igmp
igmp snooping enable
 igmp enable
[~8850_130.12-Vlanif3111]

```
### test case
```

  - name: check vlan_id not exist
    ce_multicast_igmp_enable:
      features: 'vlan'
      aftype: v4
      vlan_id: 3111
      igmp: true
      state: present

```
### test log info 
```


changed: [10.190.121.125] => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python"
    },
    "changed": true,
    "end_state": {
        "igmp_info": [
            {
                "addrFamily": "ipv4unicast",
                "proxyEnable": "false",
                "snoopingEnable": "true",
                "version": "2",
                "vlanId": "3111"
            }
        ]
    },
    "existing": {
        "igmp_info": []
    },
    "invocation": {
        "module_args": {
            "aftype": "v4",
            "features": "vlan",
            "host": "10.190.121.125",
            "igmp": "true"
            "proxy": "false",
            "ssh_keyfile": null,
            "state": "present",
            "timeout": null,
            "transport": "cli",
            "use_ssl": null,
            "username": "huawei",
            "validate_certs": null,
            "version": 2,
            "vlan_id": 3111
        }
    },
    "proposed": {
        "addrFamily": "ipv4unicast",
        "features": "vlan",
        "proxyEnable": "false",
        "snoopingEnable": "true",
        "state": "present",
        "version": 2,
        "vlanId": 3111
    },
    "updates": [
        "igmp snooping enable",
        "igmp snooping version 2",
        "undo igmp snooping proxy"
    ]
}
META: ran handlers
META: ran handlers

PLAY RECAP ******************************************************************************************************************************************************
10.190.121.125             : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```
### show on device
```
[~8850_130.12-vlan3111]di th
#
vlan 3111
 igmp snooping enable
#
return
[~8850_130.12-vlan3111]


```
### test case
```

  - name: new proxy,igmp,version
    ce_multicast_igmp_enable:
      features: 'vlan'
      aftype: v4
      vlan_id: 1
      proxy: true
      igmp: true
      version: 1

```
### test log info 
```


changed: [10.190.121.125] => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python"
    },
    "changed": true,
    "end_state": {
        "igmp_info": [
            {
                "addrFamily": "ipv4unicast",
                "proxyEnable": "true",
                "snoopingEnable": "true",
                "version": "1",
                "vlanId": "1"
            }
        ]
    },
    "existing": {
        "igmp_info": []
    },
    "invocation": {
        "module_args": {
            "aftype": "v4",
            "features": "vlan",
            "host": "10.190.121.125",
            "igmp": "true"
            "proxy": "true",
            "ssh_keyfile": null,
            "state": "present",
            "timeout": null,
            "transport": "cli",
            "use_ssl": null,
            "username": "huawei",
            "validate_certs": null,
            "version": 1,
            "vlan_id": 1
        }
    },
    "proposed": {
        "addrFamily": "ipv4unicast",
        "features": "vlan",
        "proxyEnable": "true",
        "snoopingEnable": "true",
        "state": "present",
        "version": 1,
        "vlanId": 1
    },
    "updates": [
        "igmp snooping enable",
        "igmp snooping version 1",
        "igmp snooping proxy"
    ]
}
META: ran handlers
META: ran handlers

PLAY RECAP ******************************************************************************************************************************************************
10.190.121.125             : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```
### show on device
```
[~8850_130.12-vlan1]di th
#
vlan 1
 igmp snooping enable
 igmp snooping version 1
 igmp snooping proxy
#
return
[~8850_130.12-vlan1]
```

### CODE REVIEW

File: lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py [hunk 1]
```diff
@@ -0,0 +1,557 @@
...
+    description:
+      - Distinguish between Globally Enabled IGMP or
+      - Enabled IGMP under vlanID.
+    required: true
+    type: str
+    choices: ['global','vlan']
+  vlan_id:
+    description:
+      - Virtual LAN identity.
+    required: false
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py]: required: false 
```suggestion
```

File: lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py [hunk 2]
```diff
@@ -0,0 +1,557 @@
...
+    type: int
+  igmp:
+    description:
+      - Enable Layer 2 multicast Snooping in a VLAN.
+    required: false
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py]: The same

File: lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py [hunk 3]
```diff
@@ -0,0 +1,557 @@
...
+    type: bool
+  version:
+    description:
+      - Specifies the IGMP version that can be processed.
+    required: false
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py]: the same

File: lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py [hunk 4]
```diff
@@ -0,0 +1,557 @@
...
+    default: 2
+    type: int
+  proxy:
+    description:
+      - Layer 2 multicast snooping proxy is enabled.
+    required: false
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py]: the same

File: lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py [hunk 5]
```diff
@@ -0,0 +1,557 @@
...
+    type: bool
+  state:
+    description:
+      - Specify desired state of the resource.
+    required: false
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py]: the same

File: lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py [hunk 6]
```diff
@@ -0,0 +1,557 @@
...
+    choices: ['present','absent']
+    default: present
+    type: str
+'''
+
+EXAMPLES = '''
+
+  - name: config global igmp enable
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py]:
```suggestion
  - name: configure global igmp enable
```

File: lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py [hunk 7]
```diff
@@ -0,0 +1,557 @@
...
+    ce_multicast_igmp_enable:
+      aftype: v4
+      features: 'global'
+      state: present
+
+  - name: config global igmp disable
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py]:
```suggestion
  - name: configure global igmp disable
```

File: lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py [hunk 8]
```diff
@@ -0,0 +1,557 @@
...
+    ce_multicast_igmp_enable:
+      features: 'global'
+      aftype: v4
+      state: absent
+
+  - name: config vlan igmp enable
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py]:
```suggestion
  - name: configure vlan igmp enable
```

File: lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py [hunk 9]
```diff
@@ -0,0 +1,557 @@
...
+    sample: {}
+updates:
+    description: command list sent to the device
+    returned: always
+    type: list
+    sample: ["undo igmp snooping enable",
+        "undo igmp snooping version",
+        "undo igmp snooping proxy"]
+changed:
+    description: check to see if a change was made on the device
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py]:
```suggestion
    description: check if a change was made on the device
```

File: lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py [hunk 10]
```diff
@@ -0,0 +1,557 @@
...
+        if self.features == 'vlan':
+            self.get_igmp_vlan()
+            change = self.compare_data()
+        else:
+            # sys view igmp(global)
+            self.get_igmp_global()
+            # new or edit
+            if state == 'present':
+                if not self.igmp_info_data["igmp_info"]:
+                    # self.igmp_info_data["igmp_info"] has not value
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py]:
```suggestion
                    # igmp_info_data has not igmp_info value.
```

File: lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py [hunk 11]
```diff
@@ -0,0 +1,557 @@
...
+                    change = True
+            else:
+                # delete
+                if self.igmp_info_data["igmp_info"]:
+                    # self.igmp_info_data["igmp_info"] has value
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py]:
```suggestion
                    # self.igmp_info_data has no igmp_info value
```

**Review by yuandongxx — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py (hunk 2 above)
Comment (yuandongx) [reply lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py]: Done

File: lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py (hunk 1 above)
Comment (yuandongx) [reply lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py]: Done

File: lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py (hunk 3 above)
Comment (yuandongx) [reply lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py]: Done

File: lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py (hunk 5 above)
Comment (yuandongx) [reply lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py]: Done

File: lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py (hunk 8 above)
Comment (yuandongx) [reply lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py]: Done

File: lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py (hunk 9 above)
Comment (yuandongx) [reply lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py]: Done

File: lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py (hunk 10 above)
Comment (yuandongx) [reply lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py]: Done

File: lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py (hunk 11 above)
Comment (yuandongx) [reply lib/ansible/modules/network/cloudengine/ce_multicast_igmp_enable.py]: Done

**Comment by xuxiaowei0512** [comment]
ready_for_review

**Review by yuandongxx — APPROVED** [review]

**Comment by hwDCN** [comment]
shipit

**Comment by hwDCN** [comment]
shipit

**Pull request merged**